feat: customize package details with environment variables

dopry opened the pull request:
Allow overriding package details for teams that want to fork wagtail, so they can more easily publish their fork to a repository and pip install their fork. This is useful when your team is working faster than upstream and needs to go ahead and implement features to meet their own timeline while proposing the changes upstream. Overriding packaging with environment variables will allow teams to quickly generate a forked package using CI, without needing to worry about setup.py changes ending up in PRs to upstream.

<!--
Thanks for contributing to Wagtail! 🎉

Before submitting, please review the [contributor guidelines](https://docs.wagtail.org/en/latest/contributing/index.html).
-->

_Please check the following:_

-   [ ] Do the tests still pass?[^1]
-   [ ] Does the code comply with the style guide?
    -   [ ] Run `make lint` from the Wagtail root.
-   [ ] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    -   [ ] **Please list the exact browser and operating system versions you tested**:
    -   [ ] **Please list which assistive technologies [^3] you tested**:
-   [ ] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

lb- commented:
Flagging for discussion at core team - thanks for the PR @dopry - not sure if this needs deeper discussion so just added the label needs design decision for now.

lb- commented:
Spoke about this at core team - there may be some tweaks needed for this with naming - but this is generally supported. For example `WAGTAIL_SETUPPY_NAME` kind of reads a bit odd.

@dopry  it would be great to understand a bit more of your use cases for this with specifics and examples.

I will flag this as needs review for a review further down the line.

zerolab commented:
> @dopry it would be great to understand a bit more of your use cases for this with specifics and examples.

mirroring this. 

> [..] this is useful when your team is working faster than upstream and needs to go ahead and implement features to meet their own timeline while proposing the changes upstream.

I'd like to understand more the scenario of working faster than upstream and the areas that we can improve on, as well as areas that can and should be done through 3rd party packages.

Note: #9089 and #8974 are on our list to check. Are they part of the features you mention? And are there other extension points that we could have but do not yet?

gasman commented:
> Spoke about this at core team - there may be some tweaks needed for this with naming - but this is generally supported. For example `WAGTAIL_SETUPPY_NAME` kind of reads a bit odd.

To expand on this a little - the fact that package metadata is managed by setup.py is effectively an implementation detail (for example, it's entirely possible that we might move to [pyproject.toml](https://peps.python.org/pep-0517/) in future) so it isn't really correct to reference it in the environment variable name - something like `WAGTAIL_PACKAGE_META_NAME` would be a better fit.

Actually, there's a wider point there - moving to pyproject.toml will presumably make it impossible to use environment variables in this way. While I don't think there are any immediate plans to do that, it does mean that any infrastructure you build around this pattern is liable to become defunct in the long term. Perhaps the Python world's move away from setup.py is a hint that putting variables / conditional logic in setup.py is considered an anti-pattern, and there might be some other approach (multiple config files perhaps?) that might be more aligned with where Python is heading?

dopry commented:
@lb- @zerolab  I wrote this as an example in response to this [slack thread ](https://wagtailcms.slack.com/archives/C81FGJR2S/p1663694226468549?thread_ts=1663675751.974029&cid=C81FGJR2S), but has value to me as well.   While monkey patching is working for me at the moment, it's not easy to turn those into PRs, or to merge upstream changes into the monkey patch as no merge conflicts will be raised when there are changes in the code the monkey patch is overwriting. 

@lb-  Specifically to you point of being ahead of upstream... Right now I have both monkey patches and a number of replaced models to enable SVG uploads on my current project. I have PRs against core, #8974 and #9089, that address some of that functionality. #9089 eliminates a monkey patch I had to to create because #8974 isn't merged upstream.  I've implemented SVG support on my project while it's still pending in core because my client needs it now. It's inconvenient to do it as a 3rd party package simply because the depth of monkey patching and class replacement required is tedious and hard to keep in sync with upstream changes, and they're also in my opinion extremely valuable features that belong in core.  As a  developer is a sink on my resources to have to maintain both patches against core, and my 3rd party package approaches to implementing the feature for my project.  If I could fork core and publish to my own pypi repo using circle CI, then I could maintain just a branch with SVG support in my fork and keep it in sync against upstream. In my current case my client wants SVG support now, but my my PR #8974 to enable that has been languishing in the issue queue for almost 2 months without any technical feedback. And this PR,  I wrote on a lark 2 weeks ago and haven't nagged anyone about reviewing, is getting engagement. Which isn't good or bad, just demonstrates the differing priorities and velocities of the core team and one contributor.  (side note: please review #8974 before giving this too much of your time and attention)

@gasman I've renamed the vars. In the node.js/package.json universe where env vars can't be used in package.json, it is not uncommon to overwrite the package.json prior to build as a part of the CI process to inject things like version, contributors, while leaving placeholders in the package.json itself. There are some tradeoffs to this approach. So it is possible to continue to use env vars for packaging, even if the packaging system itself doesn't natively support it.
